apps/admin_panel/assets/src/socket/connector.js

```diff
@@ -0,0 +1,161 @@
+    console.log('websocket connected.')
+    this.handleOnConnected()
+    this.drainQueue()
+    return resolve(true)
+  }
+  drainQueue = () => {
+    if (this.queueJoinChannels.length > 0) {
+      this.queueJoinChannels.forEach(channel => {
+        this.sendJoinEvent(channel)
+      })
+    }
+  }
+  addJoinedChannelToQueue = () => {
+    if (this.joinedChannels.length > 0) {
+      this.joinedChannels.forEach(channel => {
+        if (!_.includes(this.queueJoinChannels, channel)) {
+          this.queueJoinChannels.push(channel)
+        }
+      })
+      this.joinedChannels = []
+    }
+  }
+  close = () => {
+    this.handleOnDisconnected()
+    this.socket.removeEventListener('open', this.open)
+    this.socket.removeEventListener('close', this.close)
+    this.socket.removeEventListener('message', this.handleMessage)
+    clearInterval(this.heartbeat)
+    console.log('websocket disconnected.')
+    this.addJoinedChannelToQueue()
+    this.reconnect()
+  }
+  on (event, handler) {
+    switch (event) {
+      case 'connected':
+        return (this.handleOnConnected = handler)
+      case 'disconnected':
+        return (this.handleOnDisconnected = handler)
+    }
+  }
+  sendHeartbeatEvent = () => {
+    const payload = JSON.stringify({
+      topic: 'phoenix',
+      event: 'heartbeat',
+      ref: '1',
+      data: {}
+    })
+    this.socket.send(payload)
+  }
+  async reconnect () {
+    setTimeout(async () => {
+      console.log('reconnecting websocket...')
+      await this.connect()
+    }, 5000)
+  }
+  connect () {
+    return new Promise((resolve, reject) => {
+      const urlWithAuths = appendParams(this.url, this.params)
+      this.socket = new this.WebSocket(urlWithAuths)
+      this.socket.addEventListener('open', this.open(resolve))
+      this.socket.addEventListener('close', this.close)
+      this.socket.addEventListener('message', this.handleMessage)
+      this.heartbeat = setInterval(this.sendHeartbeatEvent, this.params.heartbeatInterval || 5000)
+    })
+  }
+  disconnect () {
+    this.socket.close()
+  }
+  isJoinEvent (message) {
+    return message.ref === '1' && message.topic !== 'phoenix'
+  }
+  isLeaveEvent (message) {
+    return message.ref === '2' && message.event !== 'phoenix'
+  }
+  handleMessage = message => {
+    const parsedMessage = JSON.parse(message.data)
+    if (parsedMessage.success) {
+      if (this.isJoinEvent(parsedMessage)) {
+        console.log('joined websocket channel:', parsedMessage.topic)
+        _.pull(this.queueJoinChannels, parsedMessage.topic)
+        if (!_.includes(this.joinedChannels, parsedMessage.topic)) {
+          this.joinedChannels.push(parsedMessage.topic)
+        }
+      } else if (this.isLeaveEvent(parsedMessage)) {
+        console.log('left websocket channel:', parsedMessage.topic)
+        _.pull(this.joinedChannels, parsedMessage.topic)
+        _.pull(this.queueJoinChannels, parsedMessage.topic)
+      } else {
+        // OTHER EVENT
```

Comment:
So we don't do anything here?

Reply:
there will be a handler that handle that, it will be in the next PR. I can add a dummy one first